README

Given {read,write}_handle methods, produce methods for strings and filenames.



Changes
Changelog for Mixin-Linewise

{{$NEXT}}

0.111     2022-12-31 21:09:24-05:00 America/New_York
        - update author information

0.110     2021-06-22 22:27:07-04:00 America/New_York
        - tweak the file test again to exclude directories; (is this going to
          turn into something more convoluted in several releases? probably)

0.109     2021-06-20 21:10:22-04:00 America/New_York
        - use "-r" test instead of "-f" test, to allow use on (say) named
          pipes (thanks, L. Alberto Giménez!)
        - update author information
        - add perl support section to docs

0.108     2015-04-08 14:56:06-04:00 America/New_York
        - first argument can be options only if there are two options

0.107     2015-04-07 20:14:41-04:00 America/New_York
        - add leading hashref arg for passing binmode to read_string,
          write_string
        - do not modify references of args passed to read_file, write_file

0.106     2014-03-10 08:08:39-04:00 America/New_York
        - avoid dumb v-string warning in dumb perl versions (thanks, Graham
          Knop, you're not dumb!)

0.105     2014-02-20 21:53:06-05:00 America/New_York
        - update Readers to, by default, use PerlIO::utf8_strict to avoid
          problems with CRLF layer on Win32

0.104     2013-12-21 15:19:00-0500 America/New_York
        - clean up the docs for changes made in 0.103

0.103     2013-12-19 19:34:04-0500 America/New_York
        [THIS MIGHT BREAK YOUR CODE]
        - the string methods once again open references to scalars, rather
          than using IO::Scalar

0.102     2013-11-02 11:36:22 America/New_York
        - revert the IO::String changes found in 0.101

0.101     2013-11-02 08:00:43 America/New_York
        [THIS MIGHT BREAK YOUR CODE]
        - tweak how read_string handles non-ASCII strings; needs further
          automated testing, so be sure to run your test suites

0.100     2013-10-18 15:55:59 America/New_York (TRIAL RELEASE)
        [THIS MIGHT BREAK YOUR CODE]
        - add binmode option to set binmode on opened filehandles; this
          defaults to encoding(UTF-8); thanks, David Golden!
        - typo fix by the inimitable David Steinbrunner

0.004     2013-06-18 17:30:31 America/New_York
          repackage

0.003     2010-02-14
          add Test::More prereq

0.002     2008-06-05
          update prereqs

0.001     2008-06-05
          refactored out of Config-INI


t/lib/MLTests.pm
use strict;

package MLTests;
use Mixin::Linewise::Readers
  -readers,
  -readers => { -suffix => '_sub', method => 'read_handle_sub' },
  -readers => { -suffix => '_mul', method => 'read_handle_mul' },
  -readers => { -suffix => '_utf8', method => 'read_handle_val' },
  -readers => { -suffix => '_raw', method => 'read_handle_val', encoding => ':raw' };

sub read_handle {
  my ($self, $fh) = @_;

  my %return;
  while (<$fh>) {
    chomp;
    my ($k, $v) = split /\s*=\s*/;
    ($return{$k} ||= 0) += $v;
  };

  return \%return;
}

sub read_handle_sub {
  my ($self, $fh) = @_;

  my %return;
  while (<$fh>) {
    chomp;
    my ($k, $v) = split /\s*=\s*/;
    ($return{$k} ||= 0) -= $v;
  };

  return \%return;
}

sub read_handle_mul {
  my ($self, $fh) = @_;

  my %return;
  while (<$fh>) {
    chomp;
    my ($k, $v) = split /\s*=\s*/;
    ($return{$k} ||= 1) *= $v;
  };

  return \%return;
}

sub read_handle_val {
  my ($self, $fh) = @_;

  my %return;
  while (<$fh>) {
    chomp;
    my ($k, $v) = split /\s*=\s*/;
    $return{$k} = $v;
  };

  return \%return;
}

1;


lib/Mixin/Linewise/Writers.pm
use strict;
use warnings;
package Mixin::Linewise::Writers;
# ABSTRACT: get linewise writers for strings and filenames

use 5.008001; # PerlIO
use Carp ();
use IO::File;

use Sub::Exporter -setup => {
  exports => { map {; "write_$_" => \"_mk_write_$_" } qw(file string) },
  groups  => {
    default => [ qw(write_file write_string) ],
    writers => [ qw(write_file write_string) ],
  },
};

=head1 SYNOPSIS

  package Your::Pkg;
  use Mixin::Linewise::Writers -writers;

  sub write_handle {
    my ($self, $data, $handle) = @_;

    $handle->print("datum: $_\n") for @$data;
  }

Then:

  use Your::Pkg;

  Your::Pkg->write_file($data, $filename);

  Your::Pkg->write_string($data, $string);

  Your::Pkg->write_handle($data, $fh);

=head1 EXPORTS

C<write_file> and C<write_string> are exported by default.  Either can be
requested individually, or renamed.  They are generated by
L<Sub::Exporter|Sub::Exporter>, so consult its documentation for more
information.

Both can be generated with the option "method" which requests that a method
other than "write_handle" is called with the created IO::Handle.

If given a "binmode" option, any C<write_file> type functions will use
that as an IO layer, otherwise, the default is C<encoding(UTF-8)>.

  use Mixin::Linewise::Writers -writers => { binmode => "raw" };
  use Mixin::Linewise::Writers -writers => { binmode => "encoding(iso-8859-1)" };

=head2 write_file

  Your::Pkg->write_file($data, $filename);
  Your::Pkg->write_file($data, $options, $filename);

This method will try to open a new file with the given name.  It will then call
C<write_handle> with that handle.

An optional hash reference may be passed before C<$filename> with options.
The only valid option currently is C<binmode>, which overrides any
default set from C<use> or the built-in C<encoding(UTF-8)>.

Any arguments after C<$filename> are passed along after to C<write_handle>.

=cut

sub _mk_write_file {
  my ($self, $name, $arg) = @_;
  my $method = defined $arg->{method} ? $arg->{method} : 'write_handle';
  my $dflt_enc = defined $arg->{binmode} ? $arg->{binmode} : 'encoding(UTF-8)';

  sub {
    my ($invocant, $data, $options, $filename);
    if ( ref $_[2] eq 'HASH' ) {
      # got options before filename
      ($invocant, $data, $options, $filename) = splice @_, 0, 4;
    }
    else {
      ($invocant, $data, $filename) = splice @_, 0, 3;
    }

    my $binmode = defined $options->{binmode} ? $options->{binmode} : $dflt_enc;
    $binmode =~ s/^://; # we add it later

    # Check the file
    Carp::croak "no filename specified"           unless $filename;
    Carp::croak "'$filename' is not a plain file" if -e $filename && ! -f _;

    # Write out the file
    my $handle = IO::File->new($filename, ">:$binmode")
      or Carp::croak "couldn't write to file '$filename': $!";

    $invocant->write_handle($data, $handle, @_);
  }
}

=head2 write_string

  my $string = Your::Pkg->write_string($data);
  my $string = Your::Pkg->write_string(\%option, $data);

C<write_string> will create a new handle on the given string, then call
C<write_handle> to write to that handle, and return the resulting string.
Because handles on strings must be octet-oriented, the string B<will contain
octets>.  It will be opened in the default binmode established by importing.
(See L</EXPORTS>, above, and the options, below.)

Any arguments after C<$data> are passed along after to C<write_handle>.

Like C<write_file>, this method can take a leading hashref with one valid
argument: C<binmode>.

=cut

sub _mk_write_string {
  my ($self, $name, $arg) = @_;
  my $method = defined $arg->{method} ? $arg->{method} : 'write_handle';
  my $dflt_enc = defined $arg->{binmode} ? $arg->{binmode} : 'encoding(UTF-8)';

  sub {
    my ($opt) = @_ > 2 && ref $_[1] ? splice(@_, 1, 1) : undef;
    my ($invocant, $data) = splice @_, 0, 2;

    my $binmode = ($opt && $opt->{binmode}) ? $opt->{binmode} : $dflt_enc;
    $binmode =~ s/^://; # we add it later

    my $string = '';
    open my $handle, ">:$binmode", \$string
      or die "error opening string for output: $!";

    $invocant->write_handle($data, $handle, @_);
    close $handle or die "error closing string after output: $!";

    return $string;
  }
}

1;


lib/Mixin/Linewise/Readers.pm
use strict;
use warnings;
package Mixin::Linewise::Readers;
# ABSTRACT: get linewise readers for strings and filenames

use 5.008001; # PerlIO
use Carp ();
use IO::File;
use PerlIO::utf8_strict;

use Sub::Exporter -setup => {
  exports => { map {; "read_$_" => \"_mk_read_$_" } qw(file string) },
  groups  => {
    default => [ qw(read_file read_string) ],
    readers => [ qw(read_file read_string) ],
  },
};

=head1 SYNOPSIS

  package Your::Pkg;
  use Mixin::Linewise::Readers -readers;

  sub read_handle {
    my ($self, $handle) = @_;

    LINE: while (my $line = $handle->getline) {
      next LINE if $line =~ /^#/;

      print "non-comment: $line";
    }
  }

Then:

  use Your::Pkg;

  Your::Pkg->read_file($filename);

  Your::Pkg->read_string($string);

  Your::Pkg->read_handle($fh);

=head1 EXPORTS

C<read_file> and C<read_string> are exported by default.  Either can be
requested individually, or renamed.  They are generated by
L<Sub::Exporter|Sub::Exporter>, so consult its documentation for more
information.

Both can be generated with the option "method" which requests that a method
other than "read_handle" is called with the created IO::Handle.

If given a "binmode" option, any C<read_file> type functions will use
that as an IO layer, otherwise, the default is C<utf8_strict>.

  use Mixin::Linewise::Readers -readers => { binmode => "raw" };
  use Mixin::Linewise::Readers -readers => { binmode => "encoding(iso-8859-1)" };

=head2 read_file

  Your::Pkg->read_file($filename);
  Your::Pkg->read_file(\%options, $filename);

If generated, the C<read_file> export attempts to open the named file for
reading, and then calls C<read_handle> on the opened handle.

An optional hash reference may be passed before C<$filename> with options.
The only valid option currently is C<binmode>, which overrides any
default set from C<use> or the built-in C<utf8_strict>.

Any arguments after C<$filename> are passed along after to C<read_handle>.

=cut

sub _mk_read_file {
  my ($self, $name, $arg) = @_;

  my $method = defined $arg->{method} ? $arg->{method} : 'read_handle';
  my $dflt_enc = defined $arg->{binmode} ? $arg->{binmode} : 'utf8_strict';

  sub {
    my ($invocant, $options, $filename);
    if ( ref $_[1] eq 'HASH' ) {
      # got options before filename
      ($invocant, $options, $filename) = splice @_, 0, 3;
    }
    else {
      ($invocant, $filename) = splice @_, 0, 2;
    }

    my $binmode = defined $options->{binmode} ? $options->{binmode} : $dflt_enc;
    $binmode =~ s/^://; # we add it later

    # Check the file
    Carp::croak "no filename specified"           unless $filename;
    Carp::croak "file '$filename' does not exist" unless -e $filename;
    Carp::croak "'$filename' is not readable"     unless -r _ && ! -d _;

    my $handle = IO::File->new($filename, "<:$binmode")
      or Carp::croak "couldn't read file '$filename': $!";

    $invocant->$method($handle, @_);
  }
}

=head2 read_string

  Your::Pkg->read_string($string);
  Your::Pkg->read_string(\%option, $string);

If generated, the C<read_string> creates a handle on the given string, and
then calls C<read_handle> on the opened handle.  Because handles on strings
must be octet-oriented, the string B<must contain octets>.  It will be opened
in the default binmode established by importing.  (See L</EXPORTS>, above.)

Any arguments after C<$string> are passed along after to C<read_handle>.

Like C<read_file>, this method can take a leading hashref with one valid
argument: C<binmode>.

=cut

sub _mk_read_string {
  my ($self, $name, $arg) = @_;

  my $method = defined $arg->{method} ? $arg->{method} : 'read_handle';
  my $dflt_enc = defined $arg->{binmode} ? $arg->{binmode} : 'utf8_strict';

  sub {
    my ($opt) = @_ > 2 && ref $_[1] ? splice(@_, 1, 1) : undef;
    my ($invocant, $string) = splice @_, 0, 2;

    my $binmode = ($opt && $opt->{binmode}) ? $opt->{binmode} : $dflt_enc;
    $binmode =~ s/^://; # we add it later

    Carp::croak "no string provided" unless defined $string;

    open my $handle, "<:$binmode", \$string
      or die "error opening string for reading: $!";

    $invocant->$method($handle, @_);
  }
}

1;
